File: 23. Merge k Sorted Lists/23. Merge k Sorted Lists.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode[]} lists
# @return {ListNode}
def merge_k_lists(lists)
    return nil if lists.empty?
    
    k = lists.length
    interval = 1
    while interval < k
        (0..k - interval).step(interval * 2) do |i|
            lists[i] = merge_two_lists(lists[i], lists[i + interval])
        end
        interval *= 2
    end
    
    lists[0]
end

def merge_two_lists(l1, l2)
    return l2 if l1.nil?
    return l1 if l2.nil?
    
    if l1.val <= l2.val
        l1.next = merge_two_lists(l1.next, l2)
        l1
    else
        l2.next = merge_two_lists(l1, l2.next)
        l2
    end
end

File: 23. Merge k Sorted Lists/23. Merge k Sorted Lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        if (lists.empty()) return nullptr;
        
        int k = lists.size();
        for (int interval = 1; interval < k; interval *= 2) {
            for (int i = 0; i < k - interval; i += interval * 2) {
                lists[i] = mergeTwoLists(lists[i], lists[i + interval]);
            }
        }
        
        return lists[0];
    }
    
private:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        if (!l1) return l2;
        if (!l2) return l1;
        
        if (l1->val <= l2->val) {
            l1->next = mergeTwoLists(l1->next, l2);
            return l1;
        } else {
            l2->next = mergeTwoLists(l1, l2->next);
            return l2;
        }
    }
};